pkg/metrics/metrics.go

```diff
@@ -0,0 +1,110 @@
+package metrics
+
+import (
+	"net/http"
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/rancher/apiserver/pkg/apierror"
+)
+
+type MetricLogger struct {
+	Resource string
+	Method   string
+	ID       string
+	Path     string
+}
+
+var prometheusMetrics = false
+
+const (
+	resourceLabel = "resource"
+	methodLabel   = "method"
+	codeLabel     = "code"
+	idLabel       = "id"
+	pathLabel     = "path"
+)
+
+var (
+	// https://prometheus.io/docs/practices/instrumentation/#use-labels explains logic of having 1 total_requests
+	// counter with code label vs a counter for each code
+
+	ProxyTotalResponses = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Subsystem: "k8s_proxy",
+			Name:      "total_requests",
+			Help:      "Total count API requests",
+		},
+		[]string{resourceLabel, methodLabel, codeLabel, idLabel, pathLabel},
+	)
+	K8sClientResponseTime = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Subsystem: "k8s_proxy",
+			Name:      "client_request_time",
+			Help:      "Request times in ms for k8s client from proxy store",
+		},
+		[]string{resourceLabel, methodLabel, codeLabel, idLabel, pathLabel})
+	ProxyStoreResponseTime = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Subsystem: "k8s_proxy",
+			Name:      "store_request_time",
+			Help:      "Request times in ms for k8s proxy store",
+		},
+		[]string{resourceLabel, methodLabel, codeLabel, idLabel, pathLabel})
+)
+
+func (m MetricLogger) IncTotalResponses(err error) {
+	if prometheusMetrics {
+		ProxyTotalResponses.With(
+			prometheus.Labels{
+				resourceLabel: m.Resource,
+				methodLabel:   m.Method,
+				codeLabel:     m.getAPIErrorCode(err),
+				idLabel:       m.ID,
+				pathLabel:     m.Path,
+			},
+		).Inc()
+	}
+}
+
+func (m MetricLogger) SetK8sClientResponseTime(err error, val float64) {
+	if prometheusMetrics {
+		K8sClientResponseTime.With(
+			prometheus.Labels{
+				resourceLabel: m.Resource,
+				methodLabel:   m.Method,
+				codeLabel:     m.getAPIErrorCode(err),
+				idLabel:       m.ID,
+				pathLabel:     m.Path,
+			},
+		).Observe(val)
+	}
+}
+
+func (m MetricLogger) SetProxyStoreResponseTime(err error, val float64) {
```

Comment:
nit: the names of these histogram methods are confusing because prometheus gauge metrics use `.Set(float64)`

these would be more informative as `ObserveK8sClientResponseTime` and `ObserveProxyStoreResponseTime`